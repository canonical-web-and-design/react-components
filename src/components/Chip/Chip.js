import React from "react";
import PropTypes from "prop-types";

import "./Chip.scss";

const Chip = ({ value, lead = "", onDismiss, selected }) => {
  return (
    <div className="p-chip" aria-pressed={selected} role="button">
      <span className="p-chip__lead">
        {lead ? `${lead.toUpperCase()}: ` : null}
      </span>
      <span className="p-chip__value">{value}</span>
      {onDismiss ? (
        <button className="p-chip__dismiss" onClick={() => onDismiss()}>
          <i className="p-icon--close">Dismiss</i>
        </button>
      ) : null}
    </div>
  );
};

Chip.propTypes = {
  value: PropTypes.string,
  lead: PropTypes.string,
  onDismiss: PropTypes.func,
};

export default Chip;
