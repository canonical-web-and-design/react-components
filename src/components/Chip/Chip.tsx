import { highlightSubString } from "../../utils";
import type { KeyboardEvent, MouseEvent } from "react";

export type Props = {
  /**
   * The lead for the chip.
   */
  lead?: string;
  /**
   * Function for handling chip div click event.
   */
  onClick?: (
    event: MouseEvent<HTMLDivElement> | { lead: string; value: string }
  ) => void;
  /**
   * Function for handling dismissing a chip.
   */
  onDismiss?: () => void;
  /**
   * Whether the chip is selected.
   */
  selected?: boolean;
  /**
   * A substring to emphasise if it is part of the chip's value,
   * e.g. "sit" => poSITive
   */
  subString?: string;
  /**
   * Whether to wrap the value in quotation marks.
   */
  quoteValue?: boolean;
  /**
   * The value of the chip.
   */
  value: string;
};

const Chip = ({
  lead = "",
  onClick,
  onDismiss,
  quoteValue,
  selected,
  subString = "",
  value,
}: Props): JSX.Element => {
  // When user tabs over chip and presses Enter or Space key, chip will trigger
  // click functionality
  const onKeyDown = (e: KeyboardEvent<HTMLDivElement>) => {
    // The " " value is what is returned for the spacebar
    if (e.key === " " || e.key === "Enter") {
      if (typeof onClick === "function") {
        onClick({ lead: lead, value: value });
      }
    }
  };

  const chipValue = highlightSubString(value, subString).text;
  return (
    <div
      className="p-chip"
      aria-pressed={selected}
      role="button"
      tabIndex={!onDismiss ? 0 : null}
      onClick={onClick}
      onKeyDown={(e) => onKeyDown(e)}
    >
      {lead && <span className="p-chip__lead">{lead.toUpperCase()}</span>}
      <span
        className="p-chip__value"
        dangerouslySetInnerHTML={{
          __html: quoteValue ? `'${chipValue}'` : chipValue,
        }}
      />

      {onDismiss ? (
        <button className="p-chip__dismiss" onClick={() => onDismiss()}>
          <i className="p-icon--close">Dismiss</i>
        </button>
      ) : null}
    </div>
  );
};

export default Chip;
