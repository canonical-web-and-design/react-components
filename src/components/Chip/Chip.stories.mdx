import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import Chip from "./Chip";

<Meta title="Chip" component={Chip} />

export const Template = (args) => <Chip {...args} />;

### Chip

This is a not an existing Vanilla component. It can be used to display a small value attached to a component.

### Props

<ArgsTable of={Chip} />

### Default

<Canvas>
  <Story
    name="Default"
    args={{
      value: "positive",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Key-value

<Canvas>
  <Story name="Lead-value">
    <Chip lead="Owner" value="Bob" />
  </Story>
</Canvas>

### Dismissible

<Canvas>
  <Story name="Dismissible">
    <Chip
      lead="Owner"
      value="Bob"
      onDismiss={() => {
        console.log("onDismiss called");
      }}
    />
  </Story>
</Canvas>
