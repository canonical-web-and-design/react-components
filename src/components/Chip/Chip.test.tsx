import { shallow, mount } from "enzyme";
import React from "react";

import Chip from "./Chip";

describe("Chip ", () => {
  it("renders default chip", () => {
    const wrapper = shallow(<Chip value="positive" />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders lead-value chip", () => {
    const wrapper = shallow(<Chip lead="Owner" value="Bob" />);
    expect(wrapper).toMatchSnapshot();
  });

  it("displays the value", () => {
    const wrapper = mount(<Chip value="positive" />);
    expect(wrapper.find(".p-chip").text()).toBe("positive");
  });

  it("displays the lead-value", () => {
    const wrapper = mount(<Chip lead="Owner" value="Bob" />);
    expect(wrapper.find(".p-chip").text()).toBe("OWNERBob"); // Colon and space added with CSS
  });

  it("displays the dismiss action", () => {
    const onDismiss = jest.fn();
    const wrapper = mount(
      <Chip lead="Owner" value="Bob" onDismiss={onDismiss} />
    );
    expect(wrapper.find(".p-chip__dismiss").exists()).toBe(true);
  });

  it("calls onDismiss when clicked", () => {
    const onDismiss = jest.fn();
    const wrapper = mount(
      <Chip lead="Owner" value="Bob" onDismiss={onDismiss} />
    );
    const dismissButton = wrapper.find(".p-chip__dismiss");
    dismissButton.simulate("click");
    expect(onDismiss).toHaveBeenCalledTimes(1);
  });
});
