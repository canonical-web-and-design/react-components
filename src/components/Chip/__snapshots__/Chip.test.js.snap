// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip  renders default chip 1`] = `
<div
  className="p-chip"
>
  positive
</div>
`;

exports[`Chip  renders lead-value chip 1`] = `
<div
  className="p-chip"
>
  OWNER: 
  Bob
</div>
`;
